http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcRequestAuthenticator.java

```diff
@@ -461,4 +544,124 @@ private void addScopes(String scopes, Set<String> allScopes) {
+        jwtClaims.setClaim(CLIENT_ID, deployment.getResourceName());
+
+        // sign JWT first before encrypting
+        JsonWebSignature signedRequest = signRequest(jwtClaims);
+
+        // Encrypting optional
+        if (deployment.getRequestObjectEncryptionAlgValue() != null && !deployment.getRequestObjectEncryptionAlgValue().isEmpty() &&
+            deployment.getRequestObjectEncryptionEncValue() != null && !deployment.getRequestObjectEncryptionEncValue().isEmpty()) {
+            return encryptRequest(signedRequest).getCompactSerialization();
+        } else {
+            return signedRequest.getCompactSerialization();
+        }
+    }
+
+    private static KeyPair getkeyPair(OidcClientConfiguration deployment) throws IOException {
+        if (deployment.getRequestObjectSigningKeyStoreFile().contains(PROTOCOL_CLASSPATH)) {
+            deployment.setRequestObjectSigningKeyStoreFile(deployment.getRequestObjectSigningKeyStoreFile().replace(PROTOCOL_CLASSPATH, Objects.requireNonNull(Thread.currentThread().getContextClassLoader().getResource("")).getPath()));
+        }
+        if (!deployment.getRequestObjectSigningAlgorithm().equals(NONE) && deployment.getRequestObjectSigningKeyStoreFile() == null){
+            throw log.invalidKeyStoreConfiguration();
+        } else {
+            return JWTSigningUtils.loadKeyPairFromKeyStore(deployment.getRequestObjectSigningKeyStoreFile(),
+                    deployment.getRequestObjectSigningKeyStorePassword(), deployment.getRequestObjectSigningKeyPassword(),
+                    deployment.getRequestObjectSigningKeyAlias(), deployment.getRequestObjectSigningKeyStoreType());
+        }
+    }
+
+    public JsonWebSignature signRequest(JwtClaims jwtClaims) throws IOException, JoseException {
+        JsonWebSignature jsonWebSignature = new JsonWebSignature();
+        jsonWebSignature.setPayload(jwtClaims.toJson());
+
+        if (!deployment.getRequestObjectSigningAlgValuesSupported().contains(deployment.getRequestObjectSigningAlgorithm())) {
+            throw log.invalidRequestObjectSignatureAlgorithm();
+        } else {
+            if (deployment.getRequestObjectSigningAlgorithm().contains(NONE)) { //unsigned
+                jsonWebSignature.setAlgorithmConstraints(AlgorithmConstraints.NO_CONSTRAINTS);
+                jsonWebSignature.setAlgorithmHeaderValue(NONE);
+            } else if (deployment.getRequestObjectSigningAlgorithm().contains(HMAC_SHA256)
+                    || deployment.getRequestObjectSigningAlgorithm().contains(HMAC_SHA384)
+                    || deployment.getRequestObjectSigningAlgorithm().contains(HMAC_SHA512)) { //signed with symmetric key
+                jsonWebSignature.setAlgorithmHeaderValue(deployment.getRequestObjectSigningAlgorithm());
+                Key key = new HmacKey(deployment.getResourceCredentials().get("secret").toString().getBytes(StandardCharsets.UTF_8));   //the client secret is a shared secret between the server and the client
```

Comment:
Is there a constant we can reference instead of "secret".

Can get("secret") return null?